veps/sig-storage/dic-on-heterogeneous-cluster/dic-on-heterogeneous-cluster.md

```diff
@@ -0,0 +1,538 @@
+the `DataImportCron` CR, the CDI will create architecture specific `DataVolume` CR, for each architecture
+listed in the annotation. CDI will set the `DataVolome` CR's `spec.source.registry.platform.architecture` field, to the
+required architecture.
+
+CDI will add the `template.kubevirt.io/architecture` label to the `DataSource` CR, with the architecture of the image,
+and will set the `cdi.kubevirt.io/storage.import.datasource-name` label to the value of the `spec.managedDataSource` 
+field in the `DataImportCron` CR.
+
+CDI will need to remove any orphan `DataSource` or `VolumeSnapshot` CRs, that are not referenced by any `DataImportCron`
+CR.
+
+SSP will still need to create architecture specific templates, and will still need to set the `DataSource` name 
+in each template, to the architecture specific `DataSource` CR name.
+
+Open questions:
+1. SSP currently creates the `DataSource` CRs in some cases. We have two options here:
+   1. SSP will have to have the same logic as in CDI, when creating the architecture specific `DataSource` CRs.
+   2. SSP will no longer create the `DataSource` CRs, and will let CDI to do that.
+
+2. What is the best way for SSP to find out which architecture specific templates to create? 
+   1. SSP can use the new `spec.cluster.workloadArchitectures` field in the `SSP` CR, and create the templates
+      accordingly. 
+   
+      The problem with this approach is that if a specific `DataImportCronTemplate` object supports only
+      part of the architectures listed in the `spec.cluster.workloadArchitectures` field, there will be unused templates
+      deployed. It can be mitigated by not deploying a template, if its `DataSource` was not found.
+   
+   2. SSP can use the `cdi.kubevirt.io/dict.architectures` annotation in each `DataImportCronTemplate`, and create the
+      templates accordingly.
+   
+      The problem with this approach is that SSP will need to implement the logic to find out
+      which `DataImportCronTemplate` objects is related to each template. SSP can use the `spec.managedDataSource` field
+      in the `DataImportCronTemplate` object, to find the templates using the same `DataSource` name. It may be too 
+      complex to implement this solution.
+      > **Important**: this alternative will require no API changes in SSP, as the `spec.cluster.workloadArchitectures`
+        field is not needed.
+
+pros:
+
+- no changes are required from SSP, except for adopting the new CDI API, and changes regarding the VM templates.
+- The development of the changes in SSP and in HCO is not dependent on CDI.
+- If the 2nd option in the 2nd open question is chosen, the development of the changes in HCO is not dependent on SSP,
+  and each component can be developed and released independently.
+- In deployments without SSP, this solution will work as well, if the user creates the `DataImportCron` CRs manually.
+
+cons:
+
+- template handling in SSP seems to be more complex.
```

Comment:
I don't think so, we can create templates without boot sources, that's fine.

I prefer this alternative, it requires less changes in SSP and as you pointed out it should work well in deployments without SSP too.

Reply:
@akalenyu - could you please comment on this?

Reply:
> AFAIK, there are CDI users, where it deployed w/o SSP/HCO, that do use DataImportCron. Shouldn't we support heterogeneous clusters for these users as well?

once registry DVs support specifying OCI platform fields
https://github.com/opencontainers/image-spec/blob/c05acf7eb327dae4704a4efe01253a0e60af6b34/specs-go/v1/descriptor.go#L53
the building blocks for heterogeneous clusters do exist, so why would it not be supported? 

I agree with Adam that this overlaying logic would not belong in sort of a "core" project

Reply:
I lost track a bit, so we're back to creating multiple DSes for each architecture and SSP will fill in the DS names into the Templates it creates? That would make sense to me as well.

Reply:
Creating one DICT by HCO ==> multiple DICs by SSP